reindex accessor refactor + mgmt command to populate form.server_modified_on

snopoke opened the pull request:
See also https://github.com/dimagi/commcare-hq/pull/18581

snopoke committed:
simplify reindex accessors to only require last doc PK

snopoke committed:
refactor reindex accessors to only use primary key

and convert to django queries

snopoke committed:
don't filter by id unless necessary

snopoke committed:
allow reindexers to return ids only

snopoke committed:
mark tests as use sql

snopoke committed:
add migration for populating xform.server_modified_on

snopoke committed:
update doc

snopoke committed:
fix tests

emord commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -73,14 +76,24 @@ def iter_all_rows(reindex_accessor):
+
+def iter_all_ids(reindex_accessor):
+    for db_alias in reindex_accessor.sql_db_aliases:
+        docs = reindex_accessor.get_doc_ids(db_alias)
```
Should this be `get_docs`?  Looks like you're using both get_doc_ids nad get_docs here

snopoke replied:
the other one should have been `get_doc_ids`

emord commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -234,46 +248,72 @@ def doc_to_json(self, doc):
+        :param from_db: The DB alias to query
+        :param last_doc_pk: The primary key of the last doc from the previous batch
+        :param limit: Desired batch size
+        :return: List of dict(doc_id=<document ID>, primary_key=<doc primary key>)
```
generator instaed of list. also maybe this could be a named tuple?

emord commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -234,46 +248,72 @@ def doc_to_json(self, doc):
-                return int(fetchone_as_namedtuple(cursor).reltuples)
+        query = self.query(from_db, for_count=True)
+        sql, params = query.query.sql_with_params()
+        explain_query = 'EXPLAIN {}'.format(sql)
```
Why does this explain the query?

snopoke replied:
It does and explain and then parses the row count out of it which is MUCH faster than doing a count query. Since these are only ever used for approximate counts it's ok if they aren't 100% accurate.

To be safe I've renamed the method to `get_approximate_doc_count`

emord commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -300,38 +340,15 @@ def get_doc(self, doc_id):
+        if not for_count:
+            # don't inlucde in count query since the query planner can't account for it
+            # hack for django: 'state & DELETED = 0' so 'state = state + state & DELETED'
+            filters.append(Q(state=F('state').bitand(XFormInstanceSQL.DELETED) + F('state')))
```
Does this just mean state != DELETED ? in that case would `~Q(state=DELETED)` be more obvious?

snopoke replied:
Unfortunately not: https://github.com/dimagi/commcare-hq/blob/6ae2006b2a823c4733054de0fc2de17107f26617/corehq/form_processor/models.py#L283-L286

emord replied:
Ah that makes sense. Thanks

snopoke committed:
functools.reduce

snopoke committed:
call correct function

snopoke committed:
namedtuple

snopoke committed:
rename to indicate doc count isn't accurate

snopoke committed:
fix sharded test

emord reviewed (approved):
Test failure looks real
```
======================================================================
FAIL: corehq.form_processor.tests.test_reindex_accessors:ShardedCaseReindexAccessorTests.test_get_docs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/form_processor/tests/test_reindex_accessors.py", line 75, in test_get_docs
    self.assertEqual(8, len(docs))
AssertionError: 8 != 12
----------------------------------------------------------------------
```

snopoke committed:
remove sharded reindex accessors

These tests aren't really that useful since the reindex accessors
are only ever run on a single db at a time.

millerdev commented on corehq/util/doc_processor/sql.py:
```diff
@@ -48,14 +45,10 @@ def resumable_sql_model_iterator(iteration_key, reindex_accessor, chunk_size=100
 
-    def data_function(from_db, filter_value, last_id):
-        return reindex_accessor.get_docs(from_db, filter_value, last_id, limit=chunk_size)
+    def data_function(from_db, last_id):
```
Changing this function signature made it not possible to resume iterations that were in progress (ongoing blob db migration) because [`ResumableArgsProvider`](https://github.com/dimagi/commcare-hq/blob/6143fb5b3e09a969b1a50e932550bf2360a2f27b/corehq/util/pagination.py#L131-L132) uses persisted iteration state (args passed to this function) from the old code to resume iteration.

snopoke replied:
Apologies for that, I completely forgot about those migrations. we can revert this if necessary
